hs_core/hydroshare/resource.py

```diff
@@ -464,23 +466,93 @@ def create_resource(
 
-        # prepare default metadata
-        utils.prepare_resource_default_metadata(resource=resource, metadata=metadata, res_title=title)
+        if create_metadata:
+            # prepare default metadata
+            utils.prepare_resource_default_metadata(resource=resource, metadata=metadata, res_title=title)
 
-        for element in metadata:
-            # here k is the name of the element
-            # v is a dict of all element attributes/field names and field values
-            k, v = element.items()[0]
-            resource.metadata.create_element(k, **v)
+            for element in metadata:
+                # here k is the name of the element
+                # v is a dict of all element attributes/field names and field values
+                k, v = element.items()[0]
+                resource.metadata.create_element(k, **v)
 
-        for keyword in keywords:
-            resource.metadata.create_element('subject', value=keyword)
+            for keyword in keywords:
+                resource.metadata.create_element('subject', value=keyword)
 
-        hs_bagit.create_bag(resource)
+        if create_bag:
+            hs_bagit.create_bag(resource)
 
     return resource
 
 
+def create_new_version_resource(pk, user):
+    """
+    Create a new version for a resource
+    Args:
+        pk: the unique HydroShare identifier for the resource that is to be versioned.
+        user: the requesting user who requests to create a new version for the resource
+    Returns:
+        the new resource that is created as a new version for the original resource and thus obsolete the original resource
+
+    """
+    res = utils.get_resource_by_shortkey(pk)
+    if not user.uaccess.owns_resource(res):
+        raise ValidationError('Only resource owners can create new versions')
+    # create the resource without files and without creating bags first
+    new_resource = create_resource(
+        resource_type=res.resource_type,
+        owner=user,
+        title=res.metadata.title.value,
+        create_metadata=False,
+        create_bag=False
+    )
+    # newly created new resource version is private initially
+    set_to_private = False
+    if res.raccess.public:
+        set_to_private = True
+
+    # add files directly via irods backend file operation
+    utils.copy_resource_files_and_AVUs(pk, new_resource.short_id, set_to_private)
+
+    # link copied resource files to Django resource model
+    files = ResourceFile.objects.filter(object_id=res.id)
+    for n, f in enumerate(files):
+        ResourceFile.objects.create(content_object=new_resource,
+            resource_file = os.path.join('{res_id}/data/contents/{file_name}'.format(
+                            res_id=new_resource.short_id,
+                            file_name=os.path.basename(f.resource_file.name))))
+
+    # copy metadata from source resource to target new-versioned resource except three elements
+    exclude_elements = ['identifier', 'date', 'publisher']
```

Comment:
Excluding 'date' element may cause problem. Though you are creating date element type 'created' and 'modified' there could be other date types like  'valid', and 'available' which need to be copied.